src/rules/noWarningCommentsRule.ts

```diff
@@ -0,0 +1,163 @@
+                    items: {
+                        type: "string",
+                    },
+                },
+                location: {type: "string"},
+            },
+        },
+        optionExamples: [
+            true,
+            [true, {location: "anywhere"}],
+            [true, {location: "start", terms: ["todo", "fixme", "xxx"]}],
+        ],
+        type: "style",
+        typescriptOnly: false,
+    };
+    /* tslint:enable:object-literal-sort-keys */
+
+    public static COMMENT_ERROR_FACTORY = (term: string, location: "start" | "anywhere"): string =>
+        `Warning term '${term}' found ${location === "start" ? "at start of" : "in"} comment`
+
+    public apply(sourceFile: ts.SourceFile): Lint.RuleFailure[] {
+        return this.applyWithFunction(sourceFile, walk, parseOptions(this.ruleArguments[0]));
+    }
+}
+
+function parseOptions(options: IOptionsObject): Options {
+    if (options && options.location !== "start" && options.location !== "anywhere") {
+        throw new Error(`Option 'location' value is invalid: ${options.location}`);
+    }
+
+    if (options && !(options.terms instanceof Array)) {
+        throw new Error("Option 'terms' must be a string array.");
+    }
+
+    return {
+        location: options && options.location ? options.location as "start" | "anywhere" : "start",
+        terms: options && options.terms ? options.terms : ["todo", "fixme", "xxx"],
+    };
+}
+
+function checkCommentContainsWarningTerms(comment: string, reTerm: RegExpWithTerm[], rng: ts.CommentRange, ctx: Lint.WalkContext<Options>) {
+    reTerm.forEach((regexpWithTerm) => {
+        const match = regexpWithTerm.regexp.exec(comment);
+        if (match) {
+            ctx.addFailure(rng.pos, rng.end, Rule.COMMENT_ERROR_FACTORY(regexpWithTerm.term, ctx.options.location));
+        }
+    });
+}
+
+function convertToRegExp(term: string, location: "start" | "anywhere") {
+    const escaped = term.replace(/[-/\\$^*+?.()|[\]{}]/g, "\\$&");
+    let prefix;
+
+    /*
+     * If the term ends in a word character (a-z0-9_), ensure a word
+     * boundary at the end, so that substrings do not get falsely
+     * matched. eg "todo" in a string such as "mastodon".
+     * If the term ends in a non-word character, then \b won't match on
+     * the boundary to the next non-word character, which would likely
+     * be a space. For example `/\bFIX!\b/.test('FIX! blah') === false`.
+     * In these cases, use no bounding match. Same applies for the
+     * prefix, handled below.
+     */
+    const suffix = /\w$/.test(term) ? "\\b" : "";
+
+    if (location === "start") {
+        /*
+         * When matching at the start, ignore leading whitespace, and
+         * there's no need to worry about word boundaries.
+         */
+        prefix = "^\\s*";
+    } else if (/^\w/.test(term)) {
+        prefix = "\\b";
+    } else {
+        prefix = "";
+    }
+
+    return {
+        regexp: new RegExp(prefix + escaped + suffix, "i"),
```

Comment:
I'd suggest wrapping `escaped` in parens to make it a capture group. that way you don't need to pass `term` around. you just use whatever the regexp matches.

In addition I would concat all terms to one big regular expression. And that's the only option that is passed to `walk`.
The return value of this function would then look like this:
```ts
/^\s*(todo)\b|^\s*(fixme)\b/i
```

Reply:
Why concat them before calling `walk`?  It seems a bit unnecessarily complicated.

Reply:
There is no requirement to do this. I thought maybe the regex engine can optimize it.
If you want to report multiple errors for different terms, you even need these as separate regular expressions.